src/os.zig

```diff
@@ -0,0 +1,371 @@
+                                // We got hit by a signal, let's just try again.
+                                continue;
+                            },
+                            .IO => {
+                                // I/O error.
+                                //
+                                // Perhaps trying again later will work?
+                                return error.TryAgainLater;
+                            },
+                            .MFILE => {
+                                // The maximum number (OPEN_MAX) of files was open already in the
+                                // calling process.
+                                //
+                                // Perhaps trying again later will work?
+                                return error.TryAgainLater;
+                            },
+                            .NFILE => {
+                                // The maximum number of files was open already in the system.
+                                //
+                                // Perhaps trying again later will work?
+                                return error.TryAgainLater;
+                            },
+                            .NOMEM, .RANGE => {
+                                // ENOMEM -- Insufficient memory to allocate passwd structure.
+                                // ERANGE -- Insufficient buffer space supplied.
+                                buffer_size *= buf_size_gain;
+                                temp_buf = try alloc.realloc(temp_buf, buffer_size);
+                                continue;
+                            },
+                            else => {
+                                // 0 or ENOENT or ESRCH or EBADF or EPERM or ... The given name or
+                                // uid was not found -- there's really no point in trying again.
+                                break;
+                            },
+                        }
+                    }
+
+                    return error.FailedToFindHomeDir;
+                }
+            };
+
+            if (bun.EnvVar.home.get()) |home| {
+                // The user may override $HOME with a non-absolute path. We know that's wrong.
+                if (std.fs.path.isAbsolute(home)) {
+                    return .{ ._path = .{ .unmanaged = home } };
+                }
+            }
+
+            const result = Heuristics.getpwuidr(allocator);
+            if (result != error.FailedToFindHomeDir) {
+                return result;
+            }
+
+            return error.FailedToFindHomeDir;
+        }
+    };
+};
+
+/// Windows-specific OS functionality.
+pub const Win32 = struct {
+    /// Utilities for retrieving C:\Windows, D:\Windows or whatnot. Does not support POSIX.
+    pub const WinDir = struct {
+        const Self = @This();
+
+        _path: union(enum) {
+            managed: pathlib.AutoAbsPath,
+            unmanaged: []const u8,
+        },
```

Comment:
As discussed, this should be extracted out into a separate type since it's used more than once